Move `address` out of `util`

tcharding opened the pull request:
The identifier 'util' does not convey any real information. We have a whole bunch of modules inside the `util` module.

As part of work to reduce the amount of arbitrary things in the `util` module move the `address` module to the crate root level.

- Patch 1: fixes rustfmt issues currently on master, this patch is in heaps of my open PRs at the moment
- Patch 2: Does the module move
- Patch 3: Runs the formatter, done separately so reviewers can run `rustfmt +nightly fmt` and check the diff

apoelstra commented:
Could you edit the commits so that the fuzztests build on all of them?

Kixunil commented:
This is API breaking. Should we keep old paths as deprecated reexports?

Nit: justification is kinda funny. We want to flatten `until` because it conveys no information. Being "primary type" is not relevant.

apoelstra commented:
Yeah, we should keep the `util` module for a bit just to contain deprecated exports.

tcharding commented:
Cool, no worries. Will fix justification and make this non-API breaking using re-exports. Cheers.

tcharding commented:
I added re-exports and deprecated them so as to lead downstream towards a soft upgrade path.
```rust

// We moved `address.rs`, re-export everything so as not to break the API, however deprecate the
// re-exports so folks know to upgrade sooner or later.
#[deprecated(since = "0.30.0", note = "Please use crate::address")]
mod address {
    pub use crate::address::*;
}
```

Kixunil reviewed (dismissed):
ACK c5918d21b3d4e8f0d062eac67a59c159814f47de

apoelstra commented:
53f57a8a6f901b8b975284046dd2b60cc26d09a4 still fails to compile `cargo test` in the `fuzz/` directory

tcharding commented:
The re-export `address` module was not public - face palm. `fuzz` module now uses new path `crate::address` also.

tcharding commented:
Note also, force push also includes rustdoc for re-export `address` module. I checked the HTML and the docs make sense to me.

tcharding commented:
> Could you edit the commits so that the fuzztests build on all of them?

Sorry, I missed that original comment.

tcharding commented:
FTR I'm not quite sure when to use the 'minor API change' and 'API break' labels. I get the difference but before 1.0 I don't really see the benefit of the 'minor API change' label?

apoelstra reviewed (approved):
ACK 9e7998897639ddcf115296b72fbb74ed5d4592c8

Kixunil commented:
@tcharding "minor API change" probably should've been called "API addition" instead. It can be helpful for people to quickly find which things got added in a version. But TBH I didn't work hard to tag them properly in the last release.

tcharding commented:
This PR is a bottleneck for work to do this release, marked as P-high.

sanket1729 reviewed (dismissed):
ACK 9e7998897639ddcf115296b72fbb74ed5d4592c8

Kixunil reviewed (dismissed):
ACK 9e7998897639ddcf115296b72fbb74ed5d4592c8

tcharding commented:
Wow this workflow is kinda borked
- notice ci is broken
- fix ci, do PR
- rebase this PR on ci fix
- set to draft
- wait for ci fix to merge
- rebase this PR on master
- take off draft

Any suggestions on better workflow sorely appreciated.

apoelstra commented:
I think you could avoid the "set to draft" and "edit description" steps -- if CI is red we'll know not to merge.

tcharding commented:
After rebase CI will go green, and assuming I do this in a bunch of places it may not be clear what order to merge i.e., the CI fix PR has to be merged first if we want a PR record of it, though perhaps that is not important since the commit will have all the info.

sanket1729 commented:
#1182 merged. This can be undrafted now

sanket1729 reviewed (approved):
ACK 22fd107e16804cce66ecf5b4a9e61d0532fd4e39. Reviewed range diff from 9e7998897639ddcf115296b72fbb74ed5d4592c8 which I previously ACKed.

Kixunil reviewed (approved):
ACK 22fd107e16804cce66ecf5b4a9e61d0532fd4e39

tcharding commented:
Awesome, thanks lads.

apoelstra commented:
> 

Ultimately all the info is in the git tree, yes, but I'm not sure what you mean by "merging in the right order". My thinking is that you would
* do the CI fix in a separate PR
* let that get merged
* then rebase your other PRs, on master (not on the CI-fixing PR)

apoelstra reviewed (approved):
ACK 22fd107e16804cce66ecf5b4a9e61d0532fd4e39

Kixunil commented on src/address.rs:
```diff
@@ -486,17 +474,14 @@ impl Payload {
-                script::Script::new_witness_program(version, prog)
-            }
+            Payload::WitnessProgram { version, program: ref prog } =>
+                script::Script::new_witness_program(version, prog),
```
:vomiting_face: this looks awful and confusing, any chance of changing it back?
